File: MySvelteApp.Server/internal/modules/auth/infra/token/jwt_generator.go
package token

import (
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"

	authapp "mysvelteapp/server_new/internal/modules/auth/app"
	authdomain "mysvelteapp/server_new/internal/modules/auth/domain"
)

var _ authapp.TokenGenerator = (*JWTTokenGenerator)(nil)

// JWTTokenGenerator implements TokenGenerator using github.com/golang-jwt/jwt/v5.
type JWTTokenGenerator struct {
	options    JWTOptions
	signingKey []byte
}

// NewJWTTokenGenerator validates the provided options and prepares a generator instance.
func NewJWTTokenGenerator(options JWTOptions) (*JWTTokenGenerator, error) {
	if err := options.Validate(); err != nil {
		return nil, err
	}

	keyBytes, err := DecodeKey(options.Key)
	if err != nil {
		return nil, fmt.Errorf("decode key: %w", err)
	}

	return &JWTTokenGenerator{
		options:    options,
		signingKey: keyBytes,
	}, nil
}

// GenerateToken produces a signed JWT for the supplied user entity.
func (g *JWTTokenGenerator) GenerateToken(user *authdomain.User) (string, error) {
	if user == nil {
		return "", fmt.Errorf("user must not be nil")
	}

	now := time.Now().UTC()
	expiresAt := now.Add(time.Duration(g.options.AccessTokenLifetimeHours) * time.Hour)

	claims := authClaims{
		Username: user.Username,
		NameID:   fmt.Sprintf("%d", user.ID),
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   fmt.Sprintf("%d", user.ID),
			Issuer:    g.options.Issuer,
			Audience:  []string{g.options.Audience},
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(expiresAt),
			ID:        uuid.NewString(),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)

	signedToken, err := token.SignedString(g.signingKey)
	if err != nil {
		return "", fmt.Errorf("sign token: %w", err)
	}

	return signedToken, nil
}

type authClaims struct {
	Username string `json:"name"`
	NameID   string `json:"nameid"`
	jwt.RegisteredClaims
}

File: MySvelteApp.Server/app/modules/auth/infra/token/jwt_generator.py
"""JWT token generator implementation."""

import base64
import uuid
from datetime import datetime, timedelta
from typing import Optional

import jwt

from app.modules.auth.app.ports import TokenGenerator
from app.modules.auth.domain.user import User


class JWTTokenGenerator(TokenGenerator):
    """JWT token generator using PyJWT."""

    def __init__(
        self,
        key: str,
        issuer: str,
        audience: str,
        access_token_lifetime_hours: int,
    ):
        """Initialize JWT generator."""
        self.signing_key = self._decode_key(key)
        self.issuer = issuer
        self.audience = audience
        self.access_token_lifetime_hours = access_token_lifetime_hours

        # Validate key length
        if len(self.signing_key) < 32:
            raise ValueError("jwt: key must be at least 32 bytes after decoding")

        # Validate options
        if not issuer or not issuer.strip():
            raise ValueError("jwt: issuer must be provided")
        if not audience or not audience.strip():
            raise ValueError("jwt: audience must be provided")
        if access_token_lifetime_hours < 1 or access_token_lifetime_hours > 168:
            raise ValueError("jwt: access token lifetime must be between 1 and 168 hours")

    def generate_token(self, user: User) -> str:
        """Generate a signed JWT for the user."""
        if user is None:
            raise ValueError("user must not be None")

        now = datetime.utcnow()
        expires_at = now + timedelta(hours=self.access_token_lifetime_hours)

        claims = {
            "name": user.username,
            "nameid": str(user.id) if user.id else "",
            "sub": str(user.id) if user.id else "",
            "iss": self.issuer,
            "aud": self.audience,
            "iat": int(now.timestamp()),
            "exp": int(expires_at.timestamp()),
            "jti": str(uuid.uuid4()),
        }

        token = jwt.encode(claims, self.signing_key, algorithm="HS256")
        return token

    def _decode_key(self, key: str) -> bytes:
        """Decode key from base64 or plain text format."""
        if key.startswith("base64:"):
            decoded = base64.b64decode(key[7:])
            return decoded
        return key.encode("utf-8")

